### PR TITLE
fbthrift-compiler: new brew

### PR DESCRIPTION
**Pull request by jesboat**

### CODE REVIEW

File: Library/Formula/fbthrift-compiler.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
+require "formula"
+
+class FbthriftCompiler < Formula
+  homepage "https://github.com/facebook/fbthrift"
+  url "https://github.com/facebook/fbthrift/archive/b5635e3675023979c87d8b16182402c1ffba32a4.tar.gz"
```
Comment (mistydemeo) [inline Library/Formula/fbthrift-compiler.rb]: Is there no tag for this release?

File: Library/Formula/fbthrift-compiler.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  depends_on "autoconf" => :build
+  depends_on "automake" => :build
+  depends_on "libtool" => :build
+
+  depends_on "glog"
+  depends_on "boost"
+  depends_on "folly"
+
+  def install
+    cd "thrift"
```
Comment (mistydemeo) [inline Library/Formula/fbthrift-compiler.rb]: We prefer the block form of `cd`:

```rb
cd "thrift" do
  system "autoreconf", "-i"
  # ...
end

File: Library/Formula/fbthrift-compiler.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        "--disable-dependency-tracking",
+        "--disable-silent-rules",
+        "--prefix=#{prefix}",
+        "--without-python",  # it dumps random stuff places
+        "--without-cpp"  # requires event libraries, numa, stuff we don't have
+
+    system "make", "install"
+  end
+
+  patch :p1, :DATA
```
Comment (mistydemeo) [inline Library/Formula/fbthrift-compiler.rb]: This does above the install method.

File: Library/Formula/fbthrift-compiler.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
+require "formula"
```
Comment (mistydemeo) [inline Library/Formula/fbthrift-compiler.rb]: This line isn't required anymore.

File: Library/Formula/fbthrift-compiler.rb [hunk 5]
```diff
@@ -0,0 +1,58 @@
+class FbthriftCompiler < Formula
+  homepage "https://github.com/facebook/fbthrift"
+  url "https://github.com/facebook/fbthrift/archive/v0.28.0.tar.gz"
+  version "28.0"
+  sha1 "7fed41f6d7c072b41a550fadf36ba95bdd6d9813"
```
Comment (DomT4) [inline Library/Formula/fbthrift-compiler.rb]: Needs to be `sha256`.

File: Library/Formula/fbthrift-compiler.rb [hunk 6]
```diff
@@ -0,0 +1,58 @@
+class FbthriftCompiler < Formula
+  homepage "https://github.com/facebook/fbthrift"
```
Comment (DomT4) [inline Library/Formula/fbthrift-compiler.rb]: Needs a `desc` (<80 characters) above the `homepage`.

File: Library/Formula/fbthrift-compiler.rb [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+__END__
+--- a/thrift/Makefile.am	2014-10-10 14:54:23.000000000 -0700
++++ b/thrift/Makefile.am	2014-10-10 14:54:27.000000000 -0700
+@@ -19,7 +19,7 @@
+ 
+ ACLOCAL_AMFLAGS = -I m4
+ 
+-SUBDIRS = compiler lib
++SUBDIRS = compiler
```
Comment (bfontaine) [inline Library/Formula/fbthrift-compiler.rb]: Couldn’t we do this when running `make`? E.g. with `make SUBDIRS=compiler …`?

File: Library/Formula/fbthrift-compiler.rb [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+  depends_on "autoconf" => :build
+  depends_on "automake" => :build
+  depends_on "libtool" => :build
+
+  depends_on "glog"
+  depends_on "boost"
+  depends_on "folly"
+
+  patch :p1, :DATA
```
Comment (bfontaine) [inline Library/Formula/fbthrift-compiler.rb]: Please add a comment explaining why the patch is necessary. Did you report the issue upstream?

File: Library/Formula/fbthrift-compiler.rb [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+          "--disable-debug",
+          "--disable-dependency-tracking",
+          "--disable-silent-rules",
+          "--prefix=#{prefix}",
+          "--without-python",  # it dumps random stuff places
+          "--without-cpp"  # requires event libraries, numa, stuff we don't have
+
+      system "make", "install"
+    end
+  end
```
Comment (bfontaine) [inline Library/Formula/fbthrift-compiler.rb]: Please add a test using a `test do` block.

**Comment by jesboat** [comment]
Added a comment about the patch and a test block.

> Couldn’t we do this when running make? E.g. with make SUBDIRS=compiler …?

For the top level, yeah, we could do `make SUBDIRS=compiler install`. We'd still need to patch compiler/Makefile, though, and I prefer a patch to toplevel because it makes explicit what's being changed (i.e. makes explicit that we're removing lib from SUBDIRS.)

File: Library/Formula/fbthrift-compiler.rb [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+  # 1. Thrift's configure.ac conflates the cpp runtime with the compiler. Even
+  # if we specify `--without-cpp` to skip the checks for e.g. numa which isn't
+  # available on OSX, the configure ends up forcing WITH_CPP=true anyway which
+  # tries to build the cpp runtime in lib.  We work around that by explicitly
+  # removing lib from the list of SUBDIRS to process.
+  # 2. The build process for the python chunk of the compiler doesn't really
+  # work too well. But that's ok, because it's optional, and the cpp half
+  # doesn't know how to call the python half on OSX anyway. So exclude that
+  # too.
+  patch :p1, :DATA
```
Comment (mikemcquaid) [inline Library/Formula/fbthrift-compiler.rb]: Can you please submit this patch to the upstream developers of this project and add a link to the upstream patch submission and explanation of why the patch is needed in a comment in the formula file. Thanks!

File: Library/Formula/fbthrift-compiler.rb [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    cd "thrift" do
+      system "autoreconf", "-i"
+
+      system "./configure",
+          "--disable-debug",
+          "--disable-dependency-tracking",
+          "--disable-silent-rules",
+          "--prefix=#{prefix}",
+          "--without-python", # see above
+          "--without-cpp" # see above
```
Comment (mikemcquaid) [inline Library/Formula/fbthrift-compiler.rb]: Maybe `see patch comment above`

**Comment by jesboat** [comment]
Requested changes

**Comment by jesboat** [comment]
ping

**Comment by DomT4** [comment]
Please check the audit failure before pinging us, Thanks!

```
==> brew style fbthrift-compiler
== /usr/local/Library/Formula/fbthrift-compiler.rb ==
C: 64:  5: %w-literals should be delimited by [ and ]
```

File: Library/Formula/fbthrift-compiler.rb [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+  # 2. The build process for the python chunk of the compiler doesn't really
+  # work too well. But that's ok, because it's optional, and the cpp half
+  # doesn't know how to call the python half on OSX anyway. So exclude that
+  # too.
+  # For details, see:
+  #  https://github.com/facebook/fbthrift/issues/102
+  #  https://github.com/facebook/fbthrift/issues/62
+  #  https://github.com/facebook/fbthrift/issues/50
+  #  https://github.com/facebook/fbthrift/issues/21
+  patch :p1, :DATA
```
Comment (DomT4) [inline Library/Formula/fbthrift-compiler.rb]: Any indication at all when these will be fixed upstream? We're usually fairly hesitant to carry new formulae that require patching if those patches haven't been accepted.

Comment (mikemcquaid) [inline Library/Formula/fbthrift-compiler.rb]: :+1: Please submit the patch(es) as a pull request upstream.

Comment (mikemcquaid) [inline Library/Formula/fbthrift-compiler.rb]: (and link to the pull requests instead of issues)

**Comment by DomT4** [comment]
Any update on this/the upstream situation around patching?

**Comment by DomT4** [comment]
Closing as dead. Feel free to chase up with upstream & resubmit this if still interested once there's action on some of the outstanding issues. Thanks for the submission!

**Pull request closed without merging**